File: 백준/Gold/30805. 사전 순 최대 공통 부분 수열/사전 순 최대 공통 부분 수열.py
# 사전 순 최대 공통부분 순열
import sys
input = sys.stdin.readline

a = int(input())
a_list = list(map(int,input().split()))

b = int(input())
b_list = list(map(int,input().split()))

result = []
while True:
    sort_list = sorted(a_list, reverse=True)

    max_num = -1
    
    for i in sort_list:
        if i in b_list:
            max_num = i 
            break

    if max_num == -1:
        break

    result.append(max_num)
    a_idx = a_list.index(max_num)
    b_idx = b_list.index(max_num)
        
    a_list = a_list[a_idx+1:]
    b_list = b_list[b_idx+1:]

if len(result) == 0:
    print(0)
else:
    print(len(result))
    print(*result)

File: 백준/Gold/30805. 사전 순 최대 공통 부분 수열/사전 순 최대 공통 부분 수열.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int a = Integer.parseInt(br.readLine());
        List<Integer> aList = new ArrayList<>();
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < a; i++) {
            aList.add(Integer.parseInt(st.nextToken()));
        }

        int b = Integer.parseInt(br.readLine());
        List<Integer> bList = new ArrayList<>();
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < b; i++) {
            bList.add(Integer.parseInt(st.nextToken()));
        }

        StringBuilder sb = new StringBuilder();
        int cnt = 0;

        while (true) {
            ArrayList<Integer> sortedList = new ArrayList<>(aList);
            Collections.sort(sortedList, Collections.reverseOrder());

            int maxNum = -1;
            for (int i : sortedList) {
                if (bList.contains(i)) {
                    maxNum = i;
                    break;
                }
            }

            if (maxNum == -1) {
                break;
            }

            sb.append(maxNum).append(" ");
            cnt ++;
            int aIdx = aList.indexOf(maxNum);
            int bIdx = bList.indexOf(maxNum);

            aList = aList.subList(aIdx+1, aList.size());
            bList = bList.subList(bIdx+1, bList.size());

        }

        if (cnt == 0) {
            System.out.println(0);
        } else {
            System.out.println(cnt);
            System.out.println(sb);
        }

    }
}
